File: python/llm_demo.py
import argparse
from transformers import pipeline
import time
import sys


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Demo LLM conversationnel avec GPT-2",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""Exemples:
  python llm_demo.py --prompt "Bonjour"
  python llm_demo.py --mode chat  # mode interactif
  python llm_demo.py --model distilgpt2 --max-tokens 150
        """,
    )
    parser.add_argument("--prompt", default="", help="Texte initial (ignore mode chat)")
    parser.add_argument(
        "--model",
        default="sshleifer/tiny-gpt2",
        help="Modèle Hugging Face (tiny-gpt2, distilgpt2, gpt2)",
    )
    parser.add_argument("--max-tokens", type=int, default=80, help="Tokens max générés")
    parser.add_argument(
        "--temperature", type=float, default=0.8, help="Contrôle créativité (0.1-2.0)"
    )
    parser.add_argument("--top-k", type=int, default=50, help="Top-k sampling")
    parser.add_argument("--top-p", type=float, default=0.95, help="Top-p (nucleus) sampling")
    parser.add_argument(
        "--mode",
        choices=["single", "chat"],
        default="single",
        help="single: un prompt, chat: conversation interactive",
    )
    return parser


def generate_text(generator, text: str, args) -> str:
    """Génère du texte avec timing."""
    try:
        start = time.time()
        outputs = generator(
            text,
            max_new_tokens=args.max_tokens,
            do_sample=True,
            top_k=args.top_k,
            top_p=args.top_p,
            temperature=args.temperature,
        )
        elapsed = time.time() - start
        result = outputs[0]["generated_text"]
        return result, elapsed
    except Exception as e:
        print(f"❌ Erreur de génération: {e}", file=sys.stderr)
        return None, 0


def single_mode(args):
    """Mode avec un seul prompt."""
    if not args.prompt:
        args.prompt = "Bonjour, je suis une IA intelligente et je peux"

    print(f"\n📦 Chargement du modèle '{args.model}'...")
    generator = pipeline("text-generation", model=args.model)

    print(f"✨ Génération en cours...\n")
    result, elapsed = generate_text(generator, args.prompt, args)

    if result:
        print(f"📝 Prompt:\n{args.prompt}")
        print(f"\n🤖 Réponse:\n{result}")
        print(f"\n⏱️  Temps: {elapsed:.2f}s | Tokens max: {args.max_tokens} | Temp: {args.temperature}")


def chat_mode(args):
    """Mode conversation multi-tour avec historique."""
    print(f"\n📦 Chargement du modèle '{args.model}'...")
    generator = pipeline("text-generation", model=args.model)

    history = []
    token_count = 0
    start_time = time.time()

    print(
        "\n💬 Mode conversation (tapez 'quit' pour quitter, 'clear' pour réinitialiser)\n"
    )

    while True:
        try:
            user_input = input("👤 Vous: ").strip()
        except (EOFError, KeyboardInterrupt):
            break

        if user_input.lower() == "quit":
            break
        elif user_input.lower() == "clear":
            history = []
            token_count = 0
            print("🗑️  Historique effacé.\n")
            continue
        elif not user_input:
            continue

        # Construit contexte avec historique (limite à ~500 chars pour éviter overflow)
        context = " ".join([msg for pair in history[-3:] for msg in pair])
        prompt = context + " " + user_input if context else user_input

        result, elapsed = generate_text(generator, prompt, args)

        if result:
            # Extrait la réponse nouvelle (après le prompt)
            response = result[len(prompt) :].strip() if len(result) > len(prompt) else result
            print(f"🤖 IA: {response}\n")
            history.append((user_input, response))
            token_count += args.max_tokens
        else:
            print("⚠️  Erreur lors de la génération.\n")

    # Stats finales
    total_time = time.time() - start_time
    print(f"\n📊 Stats: {len(history)} échanges, ~{token_count} tokens, {total_time:.1f}s total")


def main():
    args = build_parser().parse_args()

    if args.mode == "chat":
        chat_mode(args)
    else:
        single_mode(args)


if __name__ == "__main__":
    main()

File: julia/llm_demo.jl
#!/usr/bin/env julia

using Pkg
Pkg.activate(@__DIR__)
for pkg in ["Transformers", "ProgressMeter"]
    isnothing(Base.find_package(pkg)) && Pkg.add(pkg)
end

using Transformers
using Transformers.Basic: generate
using Transformers.HuggingFace: hgf, load_hgf_model!

function parse_args()
    prompt = ""
    model_id = "gpt2"
    max_tokens = 60
    mode = "single"  # single ou chat
    temperature = 0.8
    top_k = 50
    top_p = 0.95

    i = 1
    while i ≤ length(Base.ARGS)
        arg = Base.ARGS[i]
        if startswith(arg, "--prompt=")
            prompt = split(arg, "=", limit=2)[2]
        elseif startswith(arg, "--model=")
            model_id = split(arg, "=", limit=2)[2]
        elseif startswith(arg, "--max-tokens=")
            max_tokens = parse(Int, split(arg, "=", limit=2)[2])
        elseif startswith(arg, "--temperature=")
            temperature = parse(Float64, split(arg, "=", limit=2)[2])
        elseif startswith(arg, "--mode=")
            mode = split(arg, "=", limit=2)[2]
        elseif arg == "--help" || arg == "-h"
            println("""Utilisation:
  julia llm_demo.jl [options]

Options:
  --prompt=TEXT          Texte initial
  --model=ID             Modèle (défaut: gpt2)
  --max-tokens=N         Max tokens générés (défaut: 60)
  --temperature=T        Créativité 0.1-2.0 (défaut: 0.8)
  --mode=single|chat     Mode (défaut: single)
  --help                 Affiche cette aide
            """)
            exit(0)
        end
        i += 1
    end

    return (; prompt, model_id, max_tokens, mode, temperature, top_k, top_p)
end

function generate_text(model, tokenizer, text::String, args)
    """Génère du texte avec timing."""
    try
        start_time = time()
        generated = generate(
            model,
            tokenizer,
            text;
            max_new_tokens=args.max_tokens,
            temperature=args.temperature,
            top_k=args.top_k,
            top_p=args.top_p,
            do_sample=true,
        )
        elapsed = time() - start_time
        return generated, elapsed
    catch e
        println("❌ Erreur: $e")
        return nothing, 0
    end
end

function single_mode(args, model, tokenizer)
    """Mode avec un seul prompt."""
    prompt = isempty(args.prompt) ? "Bonjour, je suis une IA intelligente" : args.prompt
    
    println("✨ Génération en cours...\n")
    result, elapsed = generate_text(model, tokenizer, prompt, args)

    if !isnothing(result)
        println("📝 Prompt:")
        println(prompt)
        println("\n🤖 Réponse:")
        println(result)
        println("\n⏱️  Temps: $(round(elapsed, digits=2))s | Tokens: $(args.max_tokens) | Temp: $(args.temperature)")
    end
end

function chat_mode(args, model, tokenizer)
    """Mode conversation multi-tour."""
    history = String[]
    token_count = 0
    start_time = time()

    println("\n💬 Mode conversation (quit=quitter, clear=réinitialiser)\n")

    while true
        print("👤 Vous: ")
        user_input = readline()
        
        if user_input == "quit"
            break
        elseif user_input == "clear"
            empty!(history)
            token_count = 0
            println("🗑️  Historique effacé.\n")
            continue
        elseif isempty(strip(user_input))
            continue
        end

        # Contexte avec historique (limité)
        context = ""
        if length(history) > 0
            context = join(history[end-min(2, length(history)-1):end], " ")
        end
        prompt = isempty(context) ? user_input : "$context $user_input"

        result, elapsed = generate_text(model, tokenizer, prompt, args)

        if !isnothing(result)
            # Extrait réponse nouvelle
            response = if length(result) > length(prompt)
                strip(result[length(prompt)+1:end])
            else
                result
            end
            println("🤖 IA: $response\n")
            push!(history, user_input)
            push!(history, response)
            token_count += args.max_tokens
        else
            println("⚠️  Erreur de génération.\n")
        end
    end

    # Stats
    total_time = time() - start_time
    n_exchanges = div(length(history), 2)
    println("\n📊 Stats: $n_exchanges échanges, ~$token_count tokens, $(round(total_time, digits=1))s total")
end

function main()
    args = parse_args()
    
    println("📦 Chargement modèle '$(args.model_id)'...")
    tokenizer = AutoTokenizer(args.model_id)
    config = AutoConfig(args.model_id)
    model = GPT2LMHeadModel(config)
    load_hgf_model!(model, hgf"$(args.model_id)")
    
    if args.mode == "chat"
        chat_mode(args, model, tokenizer)
    else
        single_mode(args, model, tokenizer)
    end
end

main()
